### ISSUE TITLE
Full provenance for simulations

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#912]
msprime currently has half-baked support for provenance, in that only some parameters are serialised. Ideally, we want to be able to reproduce an identical tree sequence from the provenance information. This will involve serialising big things like recombination maps (sometimes) so we should also add a ``record_provenance=True`` argument, which will turn off provenance recording entirely if False. 

There's an argument for always writing the provenance, regardless (see this discussion: https://github.com/tskit-dev/pyslim/issues/67), but this seems heavy-handed in the case of msprime. Firstly, the vast majority of tree sequences generated in msprime are never stored, so provenance is irrelevant anyway. If we're doing millions of small simulations, this will be significant. Secondly,  I think we should take the Python philosophy of making it easy to write good programs rather than try to prevent people writing bad programs. There are good reasons for not wanting to write provenance info.

We'll also want a utility function ``parse_simulate_provenance(ts, index=None)`` function which will parse out the provenance from a tree sequence at a given index (defaulting to the last provenance) and return a dictionary of arguments. We should always have:
```python
ts1 = msprime.simulate(**args)
ts2 = msprime.simulate(**parse_simulate_provenance(ts1))
assert ts1.tables == ts2.tables # Well, not quite, see below
```

(We won't be able to assert exact equality, as the provenance time stamps will be different, but this is the idea anyway.)

**Comment by jeromekelleher** [comment]
We'll also want a ``parse_mutate_provenance`` function which will do the same for the ``mutate`` function.

**Comment by jeromekelleher** [comment]
The tricky bit here is dealing with replicates, and how we manage the random generator state. On the one hand, we want to be able to reproduce the a tree sequence that was produced during a ``num_replicates`` loop. On the other hand, saving the RNG state may be tricky and expensive to do in practise. (We'd have to use [this API](https://www.gnu.org/software/gsl/doc/html/rng.html#reading-and-writing-random-number-generator-state), which would be awkward.)

I think the thing to do is to have a ``replicate_index`` in the provenance, which stores the replicate number, making it at least possible to reproduce the simulation (even if we have to do all the other simulations first). I guess we could add a ``replicate_index`` option to the ``simulate`` function, which would only return the ``k``th replicate (and not an iterator over tree sequences). It's not ideal, but it would work.

I suppose we should **not** store ``num_replicates`` then within the provenance, since this is not relevant to how a particular tree sequence was generated.

**Comment by jeromekelleher** [comment]
Closed in #914.
